base/c.jl

```diff
@@ -43,6 +43,29 @@ end
 
 typealias Coff_t FileOffset
 
+# C NUL-terminated string pointers; these can be used in ccall
+# instead of Ptr{Cchar} and Ptr{Cwchar_t}, respectively, to enforce
+# a check for embedded NUL chars in the string (to avoid silent truncation).
+immutable Cstring; p::Ptr{Cchar}; end
+immutable Cwstring; p::Ptr{Cwchar_t}; end
```

Comment:
this isn't consistent with some platform ABIs, unfortunately and will break ccall on some platforms. it must be declared `bitstype WORD_SIZE Cstring`. i realize that creating new bits types is quite opaque currently and should be better documented. one extremely limited example is https://github.com/JuliaLang/julia/blob/54d8cc7e913ffeb50d479028e372f45b52d3b048/test/core.jl#L2255-L2257

pointer.jl is a more complex example (https://github.com/JuliaLang/julia/blob/master/base/pointer.jl)

(fwiw, and iirc, the only place it will actually matter (and thus segfault) is if this is used as a return type on x86 linux)